src/circuit/circuit_builder.rs

```diff
@@ -301,6 +349,107 @@ type SchedulerEventHandler = Box<dyn Fn(&SchedulerEvent)>;
 type CircuitEventHandlers = Rc<RefCell<HashMap<String, CircuitEventHandler>>>;
 type SchedulerEventHandlers = Rc<RefCell<HashMap<String, SchedulerEventHandler>>>;
 
+/// Operator's preference to consume input data by value.
+///
+/// # Background
+///
+/// A stream in a circuit can be connected to multiple consumers.  It is therefore generally
+/// impossible to provide each consumer with an owned copy of the data without cloning it.  At the
+/// same time, many operators can be more efficient when working with owned inputs.  For instance,
+/// when computing a sum of two z-sets, if one of the input z-sets is owned we can just add values
+/// from the other z-set to it without cloning the first z-set.  If both inputs are owned then we
+/// additionally do not need to clone key/value pairs when inserting them.  Furthermore, the
+/// implementation can choose to add the contents of the smaller z-set to the larger one.
+///
+/// # Ownership-aware scheduling
+///
+/// To leverage such optimizations, we adopt the best-effort approach: operators consume streaming
+/// data by-value when possible while falling back to pass-by-reference otherwise.  In a
+/// synchronous circuit, each operator reads its input stream precisely once in each clock cycle.
+/// It is therefore possible to determine the last consumer at each clock cycle and give it the
+/// owned value from the channel.  It is furthermore possible for the scheduler to schedule
+/// operators that strongly prefer owned values last.
+///
+/// We capture ownership preferences at two levels.  First, each individual operator that consumes
+/// one or more streams exposes its preferences on a per-stream basis via an API method
+/// (e.g., `[UnaryOperator::input_preference]`).  The [`Circuit`] API allows the circuit builder
+/// to override these preferences when instantiating an operator, taking into account circuit topology
+/// and workload.  We express preference as a numeric value.
+///
+/// These preferences are associated with each edge in the circuit graph.  The schedulers we have
```

Comment:
I don't think this comment belongs here. The documentation should say what the objects do, not what the state of development is. The latter will change and this comment will become stale.
If this comment belongs somewhere is in the schedulers module, and as as // comment